frame/vamm/src/lib.rs

```diff
@@ -981,21 +1185,134 @@ pub mod pallet {
 
 	// Helper functions - low-level functionality
 	impl<T: Config> Pallet<T> {
+		fn update_vamm_twap(
+			vamm_id: VammIdOf<T>,
+			vamm_state: &mut VammStateOf<T>,
+			asset_type: AssetType,
+			now: &Option<MomentOf<T>>,
+		) -> Result<DecimalOf<T>, DispatchError> {
+			// Sanity checks
+			Self::update_twap_sanity_check(vamm_state, asset_type, None, now)?;
+
+			let asset_twap_timestamp =
+				Self::asset_twap_timestamp_from_vamm_state(vamm_state, asset_type);
+			let asset_twap = Self::asset_twap_from_vamm_state(vamm_state, asset_type);
+			let new_twap = Self::calculate_twap(
+				now,
+				asset_twap_timestamp,
+				vamm_state.funding_period,
+				Self::do_get_price(vamm_state, asset_type)?,
+				Self::balance_to_decimal(asset_twap),
+			)?;
+
+			Self::do_update_twap(vamm_id, vamm_state, asset_type, new_twap, now)
+		}
+
+		fn calculate_twap(
+			now: &Option<MomentOf<T>>,
+			last_twap_timestamp: MomentOf<T>,
+			funding_period: MomentOf<T>,
+			new_price: DecimalOf<T>,
+			old_price: DecimalOf<T>,
+		) -> Result<DecimalOf<T>, DispatchError> {
+			let now = Self::now(now);
+			let weight_now: MomentOf<T> = std::cmp::max(
+				1_u64.into(),
+				now.checked_sub(&last_twap_timestamp)
+					.ok_or(Error::<T>::AssetTwapTimestampIsMoreRecent)?,
+			);
+
+			// TODO(Cardosaum): Won't this subtraction cause a failure everytime
+			// if we don't update twap for a long period of time?  for example,
+			// if `funding_period = 1 hour`, and we pass 2 or more hours without
+			// updating the twap, doesn't it mean we will throw an error each
+			// time we try to subtract `funding_period -  weight_now`?
+			let weight_last_twap: MomentOf<T> = std::cmp::max(
+				1_u64.into(),
+				funding_period
+					.checked_sub(&weight_now)
+					.ok_or(Error::<T>::FailedToComputeLastTwapWeight)?,
+			);
```

Comment:
Maybe use `.saturating_sub` (if available) to get `0` instead of an error whenever more than the period has passed